### PR TITLE
Fix TestHiveTransactionalTable flakiness

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 1]
```diff
@@ -95,9 +97,11 @@ public void testReadFullAcid(boolean isPartitioned, BucketingType bucketingType)
             compactTableAndWait(MAJOR, tableName, hivePartitionString, Duration.valueOf("5m"));
             assertThat(query(selectFromOnePartitionsSql)).containsExactly(row(20, 3), row(23, 1));
         }
-        finally {
-            onHive().executeQuery("DROP TABLE " + tableName);
-        }
+    }
+
+    private String tableName(String testName, boolean isPartitioned, BucketingType bucketingType)
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:100]: static,
 move below, at the end of the class

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 2]
```diff
@@ -190,26 +188,24 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
...
-                .map(row -> row.get("compactionid"))
-                .collect(toUnmodifiableSet());
-
         onHive().executeQuery(format("ALTER TABLE %s %s COMPACT '%s'", tableName, partitionString, compactMode.name()));
 
-        Set<String> tableCompactionsIds = Sets.difference(getTableCompactions(compactMode, tableName)
-                .map(row -> row.get("compactionid"))
-                .collect(toUnmodifiableSet()),
-                existingCompactionsIds);
+        // Since we disabled table auto compaction we can assume that every compaction is triggered in this test
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:193]: > Simplify compaction logic in TestHiveTransactionalTabl

What does this commit change?
I would prefer for this code to be defensive, especially that it's based on assumptions and hard to test.
E.g. if we want to assume there is no compaction, we should _verify_ there initially is no compaction.
(or leave as it was)

nit: typo in cmt

Comment (wendigo) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:193]: I'll add an assertion that there are no compactions

**Comment by wendigo** [comment]
AC @findepi

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 3]
```diff
@@ -185,26 +183,26 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
...
+        // Since we disabled table auto compaction and we checked that there are no compaction to the table
+        // we can assume that every compaction from now on is triggered in this test
+        // and all compaction should complete successfully before proceeding.
+        assertEventually(timeout, () -> {
+            List<Map<String, String>> allCompactions = getTableCompactions(compactMode, tableName)
+                    .collect(toImmutableList());
 
-        assertFalse(tableCompactionsIds.isEmpty(), "tableCompactionsIds are empty");
+            long completedCompactions = allCompactions.stream()
+                    .filter(row -> row.get("state").contentEquals("succeeded"))
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:197]: contentEquals -> equals

same below

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 4]
```diff
@@ -185,26 +183,26 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
 
     private static void compactTableAndWait(CompactionMode compactMode, String tableName, String partitionString, Duration timeout)
     {
-        Set<String> existingCompactionsIds = getTableCompactions(compactMode, tableName)
-                .map(row -> row.get("compactionid"))
-                .collect(toUnmodifiableSet());
-
         onHive().executeQuery(format("ALTER TABLE %s %s COMPACT '%s'", tableName, partitionString, compactMode.name()));
+        assertEquals(getTableCompactions(compactMode, tableName).count(), 0);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:187]: that should not hold, as we just invoked a compaction

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 5]
```diff
@@ -185,26 +183,26 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
...
-            Set<String> completedCompactionsIds = getTableCompactions(compactMode, tableName)
-                    .filter(row -> row.get("state").equals("succeeded"))
-                    .map(row -> row.get("compactionid"))
-                    .collect(toUnmodifiableSet());
+            long failedCompactions = allCompactions.stream()
+                    .filter(row -> row.get("state").contentEquals("failed"))
+                    .count();
 
-            assertTrue(completedCompactionsIds.containsAll(tableCompactionsIds));
+            verify(failedCompactions == 0, "compactions failed");
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:204]: we should verify there is 1 (or fewer?) compaction

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 6]
```diff
@@ -185,26 +183,27 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
...
-                existingCompactionsIds);
-
-        assertFalse(tableCompactionsIds.isEmpty(), "tableCompactionsIds are empty");
-
+        // Since we disabled table auto compaction and we checked that there are no compaction to the table
+        // we can assume that every compaction from now on is triggered in this test
+        // and all compaction should complete successfully before proceeding.
         assertEventually(timeout, () -> {
-            Set<String> completedCompactionsIds = getTableCompactions(compactMode, tableName)
+            List<Map<String, String>> allCompactions = getTableCompactions(compactMode, tableName)
```
Comment (losipiuk) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:193]: shouldn't there be just 1 row with state == `SUCCEEDED`.

And then we can just do:
```
        Map<String, String> compaction = getOnlyElement(
                getTableCompactions(compactMode, tableName).collect(ImmutableList.toImmutableList()));
        assertEquals(compaction.get("state"), "succeeded");

```

Comment (wendigo) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:193]: Yes, there should be only one row with either state succeeded or failed

**Comment by wendigo** [comment]
Failed due to https://github.com/prestosql/presto/issues/4238

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**